featuregates/controller/pkg/feature/feature_suite_test.go

```diff
@@ -0,0 +1,565 @@
+	err = (&corev1alpha2.FeatureGate{}).SetupWebhookWithManager(k8sManager)
+	Expect(err).ToNot(HaveOccurred())
+
+	go func() {
+		defer GinkgoRecover()
+		err = k8sManager.Start(ctx)
+		Expect(err).ToNot(HaveOccurred(), "failed to run manager")
+	}()
+
+	err = createResources(cfg, dynamicClient)
+	Expect(err).ToNot(HaveOccurred())
+})
+
+var _ = AfterSuite(func() {
+	cancel()
+	By("tearing down the test environment")
+	err := testEnv.Stop()
+	Expect(err).NotTo(HaveOccurred())
+})
+
+func getTestFeatureGate() *corev1alpha2.FeatureGate {
+	randomNumber, err := rand.Int(rand.Reader, big.NewInt(100000))
+	if err != nil {
+		return nil
+	}
+
+	return &corev1alpha2.FeatureGate{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: fmt.Sprintf("featuregate-%v", randomNumber),
+		},
+		Spec: corev1alpha2.FeatureGateSpec{
+			Features: []corev1alpha2.FeatureReference{},
+		},
+		Status: corev1alpha2.FeatureGateStatus{},
+	}
+}
+
+func getTestFeature() *corev1alpha2.Feature {
+	randomNumber, err := rand.Int(rand.Reader, big.NewInt(100000))
+	if err != nil {
+		return nil
+	}
+
+	return &corev1alpha2.Feature{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: fmt.Sprintf("feature-%v", randomNumber),
+		},
+		Spec: corev1alpha2.FeatureSpec{
+			Stability: corev1alpha2.Experimental,
+		},
+	}
+}
+
+var _ = Describe("Featuregate controller", func() {
+	It("Should not activate experimental features by default", func() {
+		feature := getTestFeature()
+		Expect(k8sClient.Create(ctx, feature)).Should(Succeed())
+
+		Eventually(func() bool {
+			err := k8sClient.Get(ctx, types.NamespacedName{Name: feature.Name}, feature)
+			return err == nil
+		}, timeout, interval).Should(BeTrue())
+
+		featureGate := getTestFeatureGate()
+		featureGate.Spec.Features = append(featureGate.Spec.Features, corev1alpha2.FeatureReference{
+			Name:     feature.Name,
+			Activate: false,
+		})
+		Expect(k8sClient.Create(ctx, featureGate)).Should(Succeed())
+
+		Eventually(func() bool {
+			err := k8sClient.Get(ctx, types.NamespacedName{Name: featureGate.Name}, featureGate)
+			return err == nil
+		}, timeout, interval).Should(BeTrue())
+
+		Eventually(func() bool {
+			err := k8sClient.Get(ctx, types.NamespacedName{Name: featureGate.Name}, featureGate)
+			return err == nil && len(featureGate.Status.FeatureReferenceResults) == 1
+		}, timeout, interval).Should(BeTrue())
+
+		Expect(featureGate.Status.FeatureReferenceResults[0].Status).Should(Equal(corev1alpha2.FeatureReferenceStatus("Applied")))
+		Expect(k8sClient.Get(ctx, types.NamespacedName{Name: feature.Name}, feature)).Should(BeNil())
+		Expect(feature.Status.Activated).Should(Equal(false))
+
+		featureGate.Spec.Features[0].Activate = true
+		Expect(k8sClient.Update(ctx, featureGate)).ShouldNot(Succeed())
+
+		Expect(k8sClient.Delete(ctx, feature)).Should(BeNil())
+		Expect(k8sClient.Delete(ctx, featureGate)).Should(BeNil())
+	})
+
+	It("Should activate and de-activate experimental features when permanently voided all support guarantees", func() {
```

Comment:
can you also add a test where you try to set `PermanentlyVoidAllSupportGuarantees` to false after setting it to true initially?

Reply:
Done